harbor-client/src/fedimint_client.rs

```diff
@@ -348,6 +350,192 @@ pub(crate) async fn spawn_invoice_receive_subscription(
+                ReceiveOperationState::Expired => {
+                    error!("Payment expired");
+                    sender
+                        .send(CoreUIMsgPacket {
+                            id: Some(msg_id),
+                            msg: CoreUIMsg::ReceiveFailed("Invoice expired".to_string()),
+                        })
+                        .await
+                        .unwrap();
+
+                    if let Err(e) = storage.mark_ln_receive_as_failed(operation_id) {
+                        error!("Could not mark lightning receive as failed: {e}");
+                    }
+                    break;
+                }
+                ReceiveOperationState::Failure => {
+                    error!("Payment failed");
+                    sender
+                        .send(CoreUIMsgPacket {
+                            id: Some(msg_id),
+                            msg: CoreUIMsg::ReceiveFailed("Unexpected error".to_string()),
+                        })
+                        .await
+                        .unwrap();
+
+                    if let Err(e) = storage.mark_ln_receive_as_failed(operation_id) {
+                        error!("Could not mark lightning receive as failed: {e}");
+                    }
+                    break;
+                }
+                _ => {}
+            }
+        }
+    });
+}
+
+pub(crate) async fn spawn_lnv2_payment_subscription(
+    mut sender: Sender<CoreUIMsgPacket>,
+    client: ClientHandleArc,
+    storage: Arc<dyn DBConnection + Send + Sync>,
+    operation_id: OperationId,
+    msg_id: Uuid,
+    is_transfer: bool,
+    subscription: UpdateStreamOrOutcome<SendOperationState>,
+) {
+    info!(
+        "Spawning LNv2 payment subscription for operation id: {}",
+        operation_id.fmt_full()
+    );
+    spawn(async move {
+        let mut stream = subscription.into_stream();
+        while let Some(op_state) = stream.next().await {
+            match op_state {
+                SendOperationState::Failure => {
+                    error!("Unexpected payment error");
+                    sender
+                        .send(CoreUIMsgPacket {
+                            id: Some(msg_id),
+                            msg: if is_transfer {
+                                CoreUIMsg::TransferFailure("Unexpected failure".to_string())
+                            } else {
+                                CoreUIMsg::SendFailure("Unexpected failure".to_string())
+                            },
+                        })
+                        .await
+                        .unwrap();
+
+                    if let Err(e) = storage.mark_lightning_payment_as_failed(operation_id) {
+                        error!("Could not mark lightning payment as failed: {e}");
+                    }
+                    break;
+                }
+                SendOperationState::Refunded => {
+                    error!("Payment refunded");
+                    sender
```

Comment:
I don't think we need to create a notification for the refund, will just be confusing to the user.

Reply:
Would `Canceled` be better? This state happens anytime the payment isn't successful